Add Mac Catalyst (iOS API, Mac ABI) runtime

directhex opened the pull request:
This ties into several concurrent discussions:

https://github.com/dotnet/runtime/pull/47645
https://github.com/dotnet/runtime/issues/47517
https://github.com/dotnet/runtime/issues/47518
https://github.com/dotnet/runtime/issues/44882
https://github.com/dotnet/designs/pull/174
https://github.com/dotnet/runtime/issues/47768
https://github.com/dotnet/xamarin/issues/29

I've made a few assumptions for now:

* RIDs will be `maccatalyst-x64` and `maccatalyst-arm64`
* TFM is `net6.0-maccatalyst`
* `TargetsMacCatalyst` gets defined in the build
* `IsMacCatalyst()` and `IsMacCatalystVersionAtLeast()` get added to System.Runtime (an API change in need of review)
* `--os maccatalyst` gets passed to build.sh
* xharness support for `maccatalyst` https://github.com/dotnet/xharness/issues/435

It should work well enough right now to happily build a `maccatalyst-x64` runtime pack. `maccatalyst-arm64` requires an LLVM build, which we don't have as of today, plus a newer macOS/Xcode than my current dev machine. I know I've made some incorrect assumptions regarding ARM64 (e.g. the target iOS version for Catalyst needs to map to macOS 11.0+, but the number I'm using maps to 10.15.1)

directhex committed:
Initial support for Mac Catalyst for mono subset (incl SPC)

directhex committed:
Get host and libs.native building. Lingering type forward issue w/ libs

directhex committed:
Add support for ARM64 Mac Catalyst

directhex committed:
Add platform checks. Ref https://github.com/dotnet/runtime/issues/47768

directhex committed:
Add Mac Catalyst TFM consumption where we have iOS TFM consumption

directhex committed:
Unify around maccatalyst name, fix up build a little

directhex committed:
Partially bring up app builder.

THIS IS INCOMPLETE, THERE ARE MISSING SYMBOLS LINKING THE RUNTIME

directhex committed:
Merge remote-tracking branch 'origin/master' into catalyse

directhex committed:
Finish brining up app builder. Next, test runner.y

directhex committed:
Fix test runner. Now only xharness itself is bad

directhex committed:
re-enable all tests

directhex committed:
Bad change, revert

directhex commented:
I haven't added yml to build the new stuff yet

safern commented on src/libraries/System.Private.CoreLib/src/System/OperatingSystem.cs:
```diff
@@ -203,6 +203,22 @@ public static bool IsIOSVersionAtLeast(int major, int minor = 0, int build = 0)
+#endif
+
+        /// <summary>
+        /// Check for the tvOS version (returned by 'libobjc.get_operatingSystemVersion') with a >= version comparison. Used to guard APIs that were added in the given tvOS release.
```
why is this referring to `tvOS`?

directhex replied:
Under-corrected copy-paste!

safern commented on eng/build.sh:
```diff
@@ -262,6 +262,8 @@ while [[ $# > 0 ]]; do
         osx)
           os="OSX" ;;
+        maccatalyst)
+          os="MacCatalyst" ;;
```
should `TargetArchitecture` be set to `arm64` for `MacCatalyst`?

directhex replied:
Mac Catalyst runs on macOS platforms (x64 or ARM64)

directhex committed:
copy-paste error

safern commented on src/libraries/System.Private.CoreLib/src/System.Private.CoreLib.Shared.projitems:
```diff
@@ -24,6 +24,7 @@
     <DefineConstants Condition="'$(TargetsUnix)' == 'true'">$(DefineConstants);TARGET_UNIX</DefineConstants>
     <DefineConstants Condition="'$(TargetsWindows)' == 'true'">$(DefineConstants);TARGET_WINDOWS</DefineConstants>
     <DefineConstants Condition="'$(TargetsOSX)' == 'true'">$(DefineConstants);TARGET_OSX</DefineConstants>
+    <DefineConstants Condition="'$(TargetsMacCatalyst)' == 'true'">$(DefineConstants);TARGET_CATALYST</DefineConstants>
```
Should `TARGET_CATALYST` be renamed to `TARGET_MACCATALYST` for consistency with other places where we used the full name?

safern replied:
Also in the cmake definition.

directhex replied:
I'M NOT OPPOSED TO THAT

I used to use, like, 4 different naming conventions. Then things broke because they were inconsistent between defining & checking 🙃

safern replied:
Well I see we're kind of consistent with TARGETS_* on defines and Targets* properties on MSBuild. So if it is not too much work it might make sense and also easier to reason for people.

directhex committed:
Unify around "MacCatalyst", get rid of all uses of "Catalyst"

directhex committed:
Add some yaml for Mac Catalyst

directhex committed:
Use "correct" Catalyst OS version targets - 11.0 ARM64 & 10.15.5 x64

directhex commented:
Doesn't build with updated cmake. How exciting!

directhex committed:
Disable -Woverriding-t-option on Mac Catalyst builds

Somewhere between CMake 3.17 and 3.19.4, a decision was made to force
setting of -mmacosx-version-min when building with an OSX SDK. Leave
The relevant cmake deployment variable blank, and it just picks its
own default. You can't bypass it, as far as I can tell.

Mac Catalyst requires a -target flag, which is incompatible with
a -mmacosx-version-min flag. Sadly, our best option is to suppress
the warning/error, as the bug is in CMake not with us.

directhex commented:
Relevant CMake issue is https://gitlab.kitware.com/cmake/cmake/-/issues/20132. I have a workaround, pushing to branch

directhex committed:
Fix build failure

rolfbjarne commented on src/libraries/System.Private.CoreLib/src/System/OperatingSystem.cs:
```diff
@@ -203,6 +203,22 @@ public static bool IsIOSVersionAtLeast(int major, int minor = 0, int build = 0)
+#endif
+
+        /// <summary>
+        /// Check for the Mac Catalyst version (returned by 'libobjc.get_operatingSystemVersion') with a >= version comparison. Used to guard APIs that were added in the given tvOS release.
```
```suggestion
        /// Check for the Mac Catalyst version (returned by 'libobjc.get_operatingSystemVersion') with a >= version comparison. Used to guard APIs that were added in the given Mac Catalyst release.
```

directhex committed:
Update src/libraries/System.Private.CoreLib/src/System/OperatingSystem.cs

Co-authored-by: Rolf Bjarne Kvinge <rolf@xamarin.com>

directhex committed:
Merge remote-tracking branch 'origin/master' into catalyse

akoeplinger commented on eng/pipelines/runtime-official.yml:
```diff
@@ -98,6 +98,7 @@ stages:
       - Android_x86
       - Android_arm
       - Android_arm64
+      - MacCatalyst_x64
```
add `MacCatalyst_arm64` ?

directhex replied:
I think that's blocking on arm64 LLVM packages

directhex replied:
Blocking on a different issue, https://github.com/dotnet/runtime/pull/47891

akoeplinger commented on src/libraries/Native/Unix/CMakeLists.txt:
```diff
@@ -76,6 +76,19 @@ if(CLR_CMAKE_TARGET_BROWSER)
 endif()
 
+if(CLR_CMAKE_TARGET_MACCATALYST)
+    # -target overrides -mmacosx-version-min so suppress warning about that
```
add CMake bug here too

akoeplinger commented on src/libraries/Native/Unix/CMakeLists.txt:
```diff
@@ -188,7 +201,7 @@ endif ()
-    if(NOT CLR_CMAKE_TARGET_BROWSER AND NOT CLR_CMAKE_TARGET_IOS AND NOT CLR_CMAKE_TARGET_TVOS)
+    if(NOT CLR_CMAKE_TARGET_BROWSER AND NOT CLR_CMAKE_TARGET_MACCATALYST AND NOT CLR_CMAKE_TARGET_IOS AND NOT CLR_CMAKE_TARGET_TVOS)
         if(CLR_CMAKE_TARGET_OSX)
             add_definitions(-DTARGET_OSX)
```
this define is used for ICU and SystemNative_FSync right now, we should have a tracking item somewhere to decide what we want to do for catalyst.

akoeplinger commented on src/libraries/Native/Unix/CMakeLists.txt:
```diff
@@ -216,7 +229,7 @@ include(configure.cmake)
 
-if (NOT CLR_CMAKE_TARGET_BROWSER AND NOT CLR_CMAKE_TARGET_IOS AND NOT CLR_CMAKE_TARGET_TVOS AND NOT CLR_CMAKE_TARGET_ANDROID)
+if (NOT CLR_CMAKE_TARGET_BROWSER AND NOT CLR_CMAKE_TARGET_MACCATALYST AND NOT CLR_CMAKE_TARGET_IOS AND NOT CLR_CMAKE_TARGET_TVOS AND NOT CLR_CMAKE_TARGET_ANDROID)
     add_subdirectory(System.IO.Ports.Native)
```
same here, this might turn out to be possible on catalyst?

rolfbjarne commented on src/libraries/System.Private.CoreLib/src/System/OperatingSystem.cs:
```diff
@@ -203,6 +203,22 @@ public static bool IsIOSVersionAtLeast(int major, int minor = 0, int build = 0)
+        /// Check for the Mac Catalyst version (returned by 'libobjc.get_operatingSystemVersion') with a >= version comparison. Used to guard APIs that were added in the given Mac Catalyst release.
+        /// </summary>
+        public static bool IsMacCatalystVersionAtLeast(int major, int minor = 0, int build = 0)
+            => IsMacCatalyst() && IsOSVersionAtLeast(major, minor, build, 0);
```
Versioning on Mac Catalyst is somewhat complex, because there are two versions in effect: the macOS version the app is running (or targeting), and the equivalent iOS version of the Mac Catalyst API.

Unfortunately Apple decided to use the macOS version in some places, and the iOS version in other places.

For the Xamarin parts of Mac Catalyst (API bindings), we decided to use the iOS version for API version checks (https://github.com/xamarin/xamarin-macios/issues/10453), because this is the version that Apple uses in Objective-C/Swift, and the version they use in their documentation.

Unfortunately it's not the version their version API (`[NSProcessInfo operatingSystemVersionString]`, `[UIDevice systemVersion]`) returns.

The end result is that we had to special-case the API we expose for version checks to handle Mac Catalyst versions (https://github.com/rolfbjarne/xamarin-macios/commit/9314bb5679bb246d7926ba32bb33fd2277da00c0).

I believe `IsMacCatalystVersionAtLeast` should handle versions the same way.

Ref: https://github.com/dotnet/runtime/issues/47768#issuecomment-773585963

marek-safar replied:
I'd suggest removing this part from this PR and go over API review approval separately

directhex replied:
The changes to ref/ were already removed from the PR, and the corresponding changes to SPC are commented out. Should I fully _remove_ those changes, not just comment them?

akoeplinger replied:
commenting is fine IMHO.

akoeplinger commented on src/libraries/System.Private.CoreLib/src/System/OperatingSystem.cs:
```diff
@@ -203,6 +203,22 @@ public static bool IsIOSVersionAtLeast(int major, int minor = 0, int build = 0)
+        /// <summary>
+        /// Indicates whether the current application is running on Mac Catalyst.
+        /// </summary>
+        public static bool IsMacCatalyst() =>
```
let's comment this for now so we don't touch the public API yet

akoeplinger commented on src/mono/mono.proj:
```diff
@@ -201,6 +202,21 @@
     </ItemGroup>
+    <!-- Mac Catalyst specific options -->
+    <ItemGroup Condition="'$(TargetsMacCatalyst)' == 'true'">
+      <_MonoCPPFLAGS Include="-Wno-overriding-t-option" />
```
link cmake bug

akoeplinger commented on src/mono/mono/mini/aot-compiler.c:
```diff
@@ -10030,7 +10030,7 @@ execute_system (const char * command)
 
 	g_free (quoted_command);
-#elif defined (HAVE_SYSTEM)
+#elif defined (HAVE_SYSTEM) && !defined(MAC_CATALYST)
```
instead of this we should do `set(HAVE_SYSTEM 0)` like we do for iOS in src/mono/cmake/configure.cmake

akoeplinger commented on src/mono/mono/mini/graph.c:
```diff
@@ -198,7 +198,9 @@ mono_draw_code_cfg (MonoCompile *cfg, FILE *fp)
 void
 mono_draw_graph (MonoCompile *cfg, MonoGraphOptions draw_options)
 {
+#if defined(HAVE_SYSTEM) && !defined(MAC_CATALYST)
```
same here

akoeplinger commented on src/mono/mono/mini/helpers.c:
```diff
@@ -252,7 +252,7 @@ mono_disassemble_code (MonoCompile *cfg, guint8 *code, int size, char *id)
 #endif
 
-#ifdef HAVE_SYSTEM
+#if defined(HAVE_SYSTEM) && !defined(MAC_CATALYST)
```
and here

directhex committed:
Fix folder detection for Catalyst test apps

directhex committed:
Merge remote-tracking branch 'directhex/catalyse' into catalyse

akoeplinger commented on src/tasks/AppleAppBuilder/Templates/runtime.m:
```diff
@@ -196,7 +201,7 @@
 //%DllMap%
 }
 
-#if FORCE_INTERPRETER || FORCE_AOT || !TARGET_OS_SIMULATOR
```
I don't think we can just remove this, it'd not enable AOT on e.g. iOS device builds since they don't set FORCE_AOT.
You can replace it with `|| (!TARGET_OS_SIMULATOR && !TARGET_OS_MACCATALYST)`

directhex replied:
I removed the TARGET_OS_SIMULATOR stuff on the basis I couldn't see where we were setting it - it didn't occur to me it came from Apple headers

akoeplinger commented on src/tasks/AppleAppBuilder/Xcode.cs:
```diff
@@ -162,29 +173,64 @@ public Xcode(string target)
+                    args.Append(" -scheme \"" + Path.GetFileNameWithoutExtension(xcodePrjPath) + "\"")
+                        .Append(" -destination \"platform=macOS,arch=arm64,variant=Mac Catalyst\"")
+                        .Append(" -UseModernBuildSystem=YES")
+                        .Append(" IPHONEOS_DEPLOYMENT_TARGET=13.2");
```
for arm64 maccatalyst we'll need 14.2

akoeplinger commented on src/tasks/AppleAppBuilder/Xcode.cs:
```diff
@@ -162,29 +173,64 @@ public Xcode(string target)
+                    args.Append(" -scheme \"" + Path.GetFileNameWithoutExtension(xcodePrjPath) + "\"")
+                        .Append(" -destination \"platform=macOS,arch=x86_64,variant=Mac Catalyst\"")
+                        .Append(" -UseModernBuildSystem=YES")
+                        .Append(" IPHONEOS_DEPLOYMENT_TARGET=13.2");
```
should be 13.5

directhex commented on src/libraries/System.Private.CoreLib/src/System.Private.CoreLib.Shared.projitems:
```diff
@@ -24,6 +24,7 @@
     <DefineConstants Condition="'$(TargetsUnix)' == 'true'">$(DefineConstants);TARGET_UNIX</DefineConstants>
     <DefineConstants Condition="'$(TargetsWindows)' == 'true'">$(DefineConstants);TARGET_WINDOWS</DefineConstants>
     <DefineConstants Condition="'$(TargetsOSX)' == 'true'">$(DefineConstants);TARGET_OSX</DefineConstants>
+    <DefineConstants Condition="'$(TargetsMacCatalyst)' == 'true'">$(DefineConstants);TARGET_CATALYST</DefineConstants>
```
Hah! I was checking for all lines which included 'catalyst' without 'mac' but......... obviously in this case, both are on the same line.

directhex committed:
Address changes suggested by Alex

directhex committed:
Fix unused variable when HAVE_SYSTEM is false

directhex committed:
Fix arm64 build starting. Remaining void->int error to fix due to Werror

directhex commented:
ARM64 builds blocked on:

```
  /Users/directhex/Projects/runtime/src/mono/mono/mini/exceptions-arm64.c:552:2: warning: incompatible pointer to integer conversion assigning to '__uint64_t' (aka 'unsigned long long') from 'gpointer' (aka 'void *') [-Wint-conversion]
          UCONTEXT_REG_SET_PC (sigctx, addr);
          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  /Users/directhex/Projects/runtime/src/mono/mono/mini/../../mono/utils/mono-sigcontext.h:498:24: note: expanded from macro 'UCONTEXT_REG_SET_PC'
          UCONTEXT_REG_PC (ctx) = (val); \
                                ^ ~~~~~
  1 warning generated.
```

directhex commented:
CI is boned today, and it's not my fault

directhex commented:
ARM64 will be turned on when https://github.com/dotnet/runtime/pull/47891 is fixed and merged

directhex committed:
More fixes from Alex

directhex commented:
AppleAppBuilder needs _serious_ work for Mac Catalyst ARM64, but should be good on x64

akoeplinger commented on src/mono/cmake/configure.cmake:
```diff
@@ -163,4 +163,8 @@ if(HOST_IOS)
\ No newline at end of file
+endif()
+
+if(MAC_CATALYST)
```
this should be `HOST_MACCATALYST` to be consistent with all the other platforms, you'll need to add a section for it in https://github.com/dotnet/runtime/blob/master/src/mono/CMakeLists.txt#L186 (see my other comment)

akoeplinger commented on src/mono/mono.proj:
```diff
@@ -201,6 +202,23 @@
     </ItemGroup>
+    <!-- Mac Catalyst specific options -->
+    <ItemGroup Condition="'$(TargetsMacCatalyst)' == 'true'">
+      <_MonoCMakeArgs Include="-DMAC_CATALYST=1" />
```
the variable name is a bit too generic IMHO, I'd prefer something like `-DCMAKE_SYSTEM_VARIANT=maccatalyst` that we then use in our src/mono/CMakeLists.txt to set `HOST_MACCATALYST`

akoeplinger commented on src/mono/mono.proj:
```diff
@@ -201,6 +202,23 @@
+      <_MonoCPPFLAGS Include="-Wno-overriding-t-option" />
+      <_MonoCFlags Condition="'$(TargetArchitecture)' == 'arm64'" Include="-target arm64-apple-ios14.2-macabi" />
+      <_MonoCFlags Condition="'$(TargetArchitecture)' == 'x64'" Include="-target x86_64-apple-ios13.5-macabi" />
+      <_MonoCFlags Include="-DMAC_CATALYST" />
```
this is unused

akoeplinger commented on src/mono/mono.proj:
```diff
@@ -201,6 +202,23 @@
+      <_MonoCFLAGS Condition="'$(TargetArchitecture)' == 'arm64'" Include="-arch arm64" />
+      <_MonoCXXFlags Condition="'$(TargetArchitecture)' == 'arm64'" Include="-target arm64-apple-ios14.2-macabi" />
+      <_MonoCXXFlags Condition="'$(TargetArchitecture)' == 'x64'" Include="-target x86_64-apple-ios13.5-macabi" />
+      <_MonoCXXFlags Include="-DMAC_CATALYST" />
```
this is unused

directhex committed:
Disable more tests, pending API changes

directhex committed:
Remove retired C define

CoffeeFlux commented on src/libraries/Common/tests/TestUtilities/System/PlatformDetection.cs:
```diff
@@ -28,6 +28,7 @@ public static partial class PlatformDetection
         public static bool IsNetBSD => RuntimeInformation.IsOSPlatform(OSPlatform.Create("NETBSD"));
         public static bool IsiOS => RuntimeInformation.IsOSPlatform(OSPlatform.Create("IOS"));
         public static bool IstvOS => RuntimeInformation.IsOSPlatform(OSPlatform.Create("TVOS"));
+        public static bool IsMacCatalyst => RuntimeInformation.IsOSPlatform(OSPlatform.Create("MACCATALYST"));
```
Does this need to be commented too?

akoeplinger replied:
no, it's in test code and the OSPlatform.Create() string based API still works.

directhex committed:
More changes from Alex

directhex committed:
Merge remote-tracking branch 'origin/master' into catalyse

directhex committed:
Fixes from Zoltan and Steve for arm64

directhex committed:
Build Catalyst for arm64 too

directhex commented:
I've dispatched an official build, since my "oops it broke official builds" success rate isn't great

directhex commented:
passed on internal
